### PR TITLE
CSS: use setProperty to support !important

### PR DESCRIPTION
**Pull request by Sulokit**
### Summary ###
Resolution to issue #3713, where !important was not taken when css property was been set using $.css method.

Please review and let me know if any changes are required.

### Checklist ###

* [ ] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [ ] New tests have been added to show the fix or feature works
* [ ] Grunt build and unit tests pass locally with these changes
* [ ] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

### CODE REVIEW

File: src/css.js [hunk 1]
```diff
@@ -266,6 +266,11 @@ jQuery.extend( {
 
 				if ( isCustomProp ) {
 					style.setProperty( name, value );
+				} else if ( value.indexOf( "!important" !== -1 ) ) {
```
Comment (mgol) [inline src/css.js]: Custom properties' values may have the `!important` clause as well, this needs to be handled as well.

File: src/css.js [hunk 2]
```diff
@@ -266,6 +266,11 @@ jQuery.extend( {
...
+
+					//Added else if to resolve #3713
```
Comment (mgol) [inline src/css.js]: Comments need to start with space.

Also, I'd rephrase to:
```js
// Support !important priority (gh-3713).
```

File: src/css.js (hunk 1 above)
Comment (mgol) [reply src/css.js]: Right. Unit tests would immediately catch this mistake. :)

**Comment by Sulokit** [comment]
Thank you very much for your comments. This is my initial phase of starting my contribution to the open source community. I would need your reviews and help to understand and help the community in the better way. I've made the changes as per the comments, please review it and let me know if any other changes are needed. :). It would really be helpful. I would need help in running the unit tests. If there is any documentation of running the tests, please share it here, so that I can go through it and implement. Currently, I've tested multiple scenarios.

**Comment by Sulokit** [comment]
Also, I've already signed up for Contributor License Agreement, I'm not sure, why it's showing as pending. Should I raise a ticket for the issue to be resolved?

File: src/css.js [hunk 3]
```diff
@@ -263,8 +264,12 @@ jQuery.extend( {
 			// If a hook was provided, use that value, otherwise just set the specified value
 			if ( !hooks || !( "set" in hooks ) ||
 				( value = hooks.set( elem, value, extra ) ) !== undefined ) {
+				if ( typeof value === "string" && value.includes( "!important" ) ) {
 
-				if ( isCustomProp ) {
+					// Support !important priority (gh-3713).
+					var splitValue = value.split( "!" );
+					style.setProperty( argName,  splitValue[ 0 ], splitValue[ 1 ] );
+				} else if ( isCustomProp ) {
```
Comment (dmethvin) [inline src/css.js]: Is it valid to say that custom properties never use `!mportant`? I would sure hope you can't override CSS variables with it or some other crazy thing 😵 but I'll ask anyway.

File: src/css.js [hunk 4]
```diff
@@ -263,8 +264,12 @@ jQuery.extend( {
 			// If a hook was provided, use that value, otherwise just set the specified value
 			if ( !hooks || !( "set" in hooks ) ||
 				( value = hooks.set( elem, value, extra ) ) !== undefined ) {
+				if ( typeof value === "string" && value.includes( "!important" ) ) {
 
-				if ( isCustomProp ) {
+					// Support !important priority (gh-3713).
+					var splitValue = value.split( "!" );
+					style.setProperty( argName,  splitValue[ 0 ], splitValue[ 1 ] );
```
Comment (dmethvin) [inline src/css.js]: This may (usually *will*) leave trailing space on the value string, is that okay? Otherwise it should be trimmed.

File: src/css.js (hunk 3 above)
Comment (mgol) [reply src/css.js]: CSS vars do support “!important”. They’re basically like any other CSS prop, just with some additional features, the most important one that they’re not dropped by the engine what happens with regular unrecognized CSS props.

File: src/css.js (hunk 4 above)
Comment (mgol) [reply src/css.js]: It should not be trimmed and I’d like to see a test that verifies that. The CSS var value is everything between the colon after the property name and “!important”. E.g. for the declaration:
```css
--foo: bar !important;
```
Running `getPropertyValue("--foo") on the matches element style returns `" bar "`, with both leading & trailing spaces.

File: src/css.js [hunk 5]
```diff
@@ -263,8 +264,12 @@ jQuery.extend( {
 			// If a hook was provided, use that value, otherwise just set the specified value
 			if ( !hooks || !( "set" in hooks ) ||
 				( value = hooks.set( elem, value, extra ) ) !== undefined ) {
+				if ( typeof value === "string" && value.includes( "!important" ) ) {
```
[the remark this answers is not in the document]
Comment (Sulokit) [reply src/css.js]: Thank you for your suggestion, I've replaced it with indexOf.. :)

File: src/css.js (hunk 4 above)
Comment (Sulokit) [reply src/css.js]: @dmethvin getPropertyValue will work as usual. It won't be affected by `!important`

**Comment by timmywil** [comment]
This is looking good. However, it needs tests before we can merge. The [README](https://github.com/jquery/jquery#running-the-unit-tests) has instructions for how to run the unit tests as you develop, but you can also run all tests with `npm test`.

File: src/css.js [hunk 6]
```diff
@@ -217,6 +217,7 @@ jQuery.extend( {
 
 		// Make sure that we're working with the right name
 		var ret, type, hooks,
+			argName = name,
```
Comment (mgol) [inline src/css.js]: Why do you need this variable?

File: src/css.js [hunk 7]
```diff
@@ -263,8 +264,12 @@ jQuery.extend( {
 			// If a hook was provided, use that value, otherwise just set the specified value
 			if ( !hooks || !( "set" in hooks ) ||
 				( value = hooks.set( elem, value, extra ) ) !== undefined ) {
+				if ( typeof value === "string" && value.indexOf( "!important" ) !== -1 ) {
 
-				if ( isCustomProp ) {
+					// Support !important priority (gh-3713).
+					var splitValue = value.split( "!" );
+					style.setProperty( argName,  splitValue[ 0 ], splitValue[ 1 ] );
```
Comment (mgol) [inline src/css.js]: This will not work for properties that need to be prefixed as you're using the unprocessed `argName` variable.

**Comment by timmywil** [comment]
This has gotten old and stale. We'll need someone to take over.
